### PR TITLE
Added story and votes info, updated popup UI colors, fix scrolling overshoot and more than one movie generation issue

### PR DESCRIPTION
**Pull request by AMMARmms**
- Modified getMultimedias.js:
  - extracted story and number of votes info
  - Fixed scrolling overshoot after reaching the last element

- Updated popup.js:
  - created story and number of votes element
  - Resolved more than one movie generation issue when "randomize'' button is pressed after "scan & rand" button clicked more than once 

- Updated popup.html and popup.css:
  - Changed colors of popup UI

### CODE REVIEW

File: getMultimedias.js [hunk 1]
```diff
@@ -70,14 +75,31 @@
...
+	// const scrollAmount = document.documentElement.scrollHeight - document.documentElement.clientHeight;
     const scrollAmount = 212 * 25; // each element's height is close to 212px and to make sure we load all the images we scroll 25 elements by 25
     document.documentElement.scrollTop += scrollAmount;
-
-    const allTitles = document.querySelectorAll('.ipc-metadata-list-summary-item');
+	
+    const numOfTitlesAsInt = parseInt(numOfTitles);  
+    const maxScrollPos = ((numOfTitlesAsInt - 3) * 207) + 450; // Set max scroll position based on num of titles, avg element's height 207px and the lenghth from page top to 1st element in the list 450px 
+    // Define event listener function
+    function handleScroll() {
```
Comment (AMMARmms) [inline getMultimedias.js]: The idea here is putting max position from top in pixels that the scroll bar should not exceed, the max pos calculated based on number of  titles in the page, the average height of the title and the distance between the top and 1st movie in the page.

File: popup.js [hunk 2]
```diff
@@ -243,22 +259,27 @@ const saveWatchlistToLocal = (multimedias) => {
   );
 };
 
+const randomizeFromCacheBtn = document.querySelector("#randomizeFromSync");
+randomizeFromCacheBtn.hasEventListener = false;
 const loadWatchlistFromLocal = (randomizeFromCacheBtn, listenerCBack) => {
```
Comment (AMMARmms) [inline popup.js]: The problem in this function was that every time u click the "scan & rand" button a new event listener for the "Randomize" Button is added, so for example if u clicked "scan & rand" button 3 times and after that u clicked "Randomize" button only once, it would be considered as it was clicked 3 times, because there is 3  event listener created.
Here, I solved that problem by using event listener flag.

Comment (AlkTheOrg) [reply popup.js]: Great catch!
But I feel like our approach is wrong on solving this. What you did is like the cleanup function in a `React.useEffect` whereas here we have full control on plain JS. I think a better question would be why are we letting users click on it multiple times in the first place? I was really amateur when I first made this extension, so looks like I couldn't cover UX very nice.

I think we can just block users to click on it several times or we can just restart the process from zero once clicked again, but I don't see a use case for the latter.

**Comment by AlkTheOrg** [comment]
> Changed colors of popup UI

I am not a designer, but I'm really not a fan of the new colors and some texts are too small for a11y users. If you don't have any a11y concerns with the current colors, let's stick with the previous ones please.

File: getMultimedias.js (hunk 1 above)
Comment (AlkTheOrg) [reply getMultimedias.js]: I thought about this when I first updated the code, but we can not ensure that this is going to be covering different base font sizes of users that they set from their browsers. We can assume base px value is going to be 16px, but I don't think the effort for those calculations will worth it. In summary your `masScrollPos` calculation will be false for some users.

Let's stick with the current safe one, and one flawed approach I have currently is the `isInView` check, but I will update it.

**Comment by AlkTheOrg** [comment]
> story and number of votes info

- We need to be able to control the `story`. Right now it is rendered no matter what whereas I would like to hide it from my results.
- Number of voters is too big and unreadable because of the color you chose. It can be below the rating and in parenthesizes. Also the spacing between the story and the ratings is not enough.

image

File: popup.html [hunk 3]
```diff
@@ -22,7 +22,8 @@
         alt="logo"
         class="m-0"
       />
-
+	  
+	  <label class="creator m-2" style="color: #f5c518;">AlkTheOrg</label>
```
Comment (AlkTheOrg) [inline popup.html]: No need to credit my name. Let's stick with the old layout.

File: popup.html [hunk 4]
```diff
@@ -135,14 +136,16 @@
             type="button"
             id="scanAndRandomize"
             class="btn btn-outline-danger btn-sm"
+			style="background-color: #0074D9; color: #FFFFFF;"
           >
-            SCAN & RANDOMIZE
+            SCAN & RAND
```
Comment (AlkTheOrg) [inline popup.html]: Some people might not understand the abbreviation, let's stick with the old version.

File: getMultimedias.js (hunk 1 above)
Comment (AMMARmms) [reply getMultimedias.js]: I made another code that's much better and doesn't depend on fonts. I'll push it soon.

File: popup.js (hunk 2 above)
Comment (AMMARmms) [reply popup.js]: If we block the button, users will have no way to update their saved watchlist.

File: popup.html (hunk 4 above)
Comment (AMMARmms) [reply popup.html]: Thank you for the review. I'll stick with the old style and colors. and I'll update the story and votes element according to your feedback.

File: popup.js (hunk 2 above)
Comment (AlkTheOrg) [reply popup.js]: We should block the click while the scan is in progress and enable again after.

Comment (AMMARmms) [reply popup.js]: You didn't get me well, the problem happen when someone click scan & randomize button even if scan is finished. 
here's a video to make it more clear.   


https://github.com/AlkTheOrg/imdb-watchlist-randomizer/assets/110132520/1d1abbaf-7cda-49e6-b1f1-59e218a8b462

**Comment by AMMARmms** [comment]
@AlkTheOrg  I've made the requested changes. Can you take another look?

File: getMultimedias.js (hunk 1 above)
Comment (AlkTheOrg) [reply getMultimedias.js]: What I tried to say is that the optimal solution should not need anything related with the sizing. If the last element is in dom, then we can just scroll into it and done. In your solution you are creating an event listener and removing it with a setTimeout which is creating unnecessary complexity and hard to read.

File: popup.html [hunk 5]
```diff
@@ -54,25 +54,25 @@
...
-            <option value="1" selected>All</option>
-            <option value="2">2</option>
-            <option value="3">3</option>
-            <option value="4">4</option>
-            <option value="5">5</option>
-            <option value="6">6</option>
-            <option value="7">7</option>
-            <option value="8">8</option>
-            <option value="9">9</option>
+          <label class="" for="minIMDBRating" style="color: #333333;">Min IMDB Rating:</label>
```
Comment (AlkTheOrg) [inline popup.html:57]: I thought we decided to revert all color changes, but you have bunch of inline stylings here. Also we have bootstrap, so why do you need inline styles for white and black?

File: popup.js (hunk 2 above)
Comment (AlkTheOrg) [reply popup.js]: Then this is an extremely rare case, because if you close the popup js all event listeners will be lost right? Although I have some feedback to give you on your code, I don't think we need that logic at all. Also the main problem is my initial solution which is using the multimedia from the callback's stale data which is a very bad practice in this case, but as I said I was very amateur when I made this extension.

File: getMultimedias.js [hunk 6]
```diff
@@ -23,13 +23,16 @@
 
       const [yearElement, runtimeElement] = div.querySelectorAll('.dli-title-metadata-item');
 
-      const year = yearElement.textContent;
-      const runtime = runtimeElement.textContent;
+      const year = yearElement ? yearElement.textContent.trim() : '';
+      const runtime = runtimeElement ? runtimeElement.textContent.trim() : '';
```
Comment (AlkTheOrg) [inline getMultimedias.js]: Do you have any movie/tv-series examples where these don't exist?

File: getMultimedias.js [hunk 7]
```diff
@@ -23,13 +23,16 @@
...
 
       // const genres = getTextContentIfExists(div, ".genre");
+	  const voteCountElement = div.querySelector('.ipc-rating-star--voteCount');
+      const voteCount = voteCountElement ? voteCountElement.textContent.trim() : '';
```
Comment (AlkTheOrg) [inline getMultimedias.js]: Could we get rid of the vote count? I didn't put it intentionally as I would like it to be conditionally rendered as well, but I also don't want to pollute the inputs with the current design. Maybe in the future

File: getMultimedias.js [hunk 8]
```diff
@@ -23,13 +23,16 @@
...
       const ratingIMDB = div.querySelector('[data-testid="ratingGroup--imdb-rating"]').textContent.replace(/\u00A0/g, ' ').split(' ')[0];
       const ratingMetac = getTextContentIfExists(div, ".metacritic-score-box");
-
+	  
+      const story = div.querySelector('.ipc-html-content-inner-div').textContent;
```
Comment (AlkTheOrg) [inline getMultimedias.js]: Could we rename all `story` stuff to `plot` please?

File: popup.css [hunk 9]
```diff
@@ -101,18 +101,25 @@ border-radius: 5px;
...
+	  color: #FFFFFF; 
+       left: 8%;	  
+  }
+  
   .ratings,
   .runtime {
     position: absolute;
     top: 50%;
     transform: translateY(-50%);
+	color:  #FFFFFF;
```
Comment (AlkTheOrg) [inline popup.css]: Why is this needed? Computed color of this element is already white in the current one.

File: popup.css [hunk 10]
```diff
@@ -101,18 +101,25 @@ border-radius: 5px;
...
   }
 
   .ratings {
     list-style: none;
     padding: 0;
-    left: 12%;
+    left: 8%;
```
Comment (AlkTheOrg) [inline popup.css:114]: Looks much better now thanks!

File: popup.css [hunk 11]
```diff
@@ -149,9 +156,18 @@ border-radius: 5px;
   }
 
   .runtime {
-    right: 12%;
+    right: 8%;
```
Comment (AlkTheOrg) [inline popup.css:151]: Looks much better now thanks!

File: popup.css [hunk 12]
```diff
@@ -149,9 +156,18 @@ border-radius: 5px;
...
   }
+  .story {
+	  font-size: 15px;
+	  text-align: center;
+	  color: #FFFFFF;
+  }
+  .credits {
+	  font-size: 11px;
+	  color: #ADD8E6;
+  }
```
Comment (AlkTheOrg) [inline popup.css]:
```suggestion
  .story {
	  font-size: 14px;
  }
  .credits {
	  font-size: 12px;
	  color: #ADD8E6;
  }
```
Credits can be a lighter color.

File: popup.html [hunk 13]
```diff
@@ -22,7 +22,7 @@
         alt="logo"
         class="m-0"
       />
-
+
```
Comment (AlkTheOrg) [inline popup.html]: Unneeded blank

File: popup.html [hunk 14]
```diff
@@ -32,7 +32,7 @@
               target="_blank"
               href="https://github.com/AlkTheOrg/imdb-watchlist-randomizer/blob/main/ABOUT.md#about-the-extension"
               >About</a
-            >
+			>
```
Comment (AlkTheOrg) [inline popup.html]: Unneeded blank

File: popup.html [hunk 15]
```diff
@@ -54,25 +54,25 @@
...
         <div class="col-md-6 mb-3 mr-auto ml-4" style="width: auto">
           <label class="" for="multimediaType">Multimedia Type: </label>
-          <select class="form-control form-control-sm" id="multimediaType">
-            <option value="movie" selected>Movies Only</option>
-            <option value="series">TV Series Only</option>
-            <option value="all">All</option>
+          <select class="form-control form-control-sm" id="multimediaType" style="background-color: #DDDDDD; color: #000000; border: 1px solid #001f3f; padding: 5px; border-radius: 5px;">
+            <option value="movie" selected style="background-color: #FFFFFF; color: #000000;">Movies Only</option>
+            <option value="series" style="background-color: #FFFFFF; color: #000000;">TV Series Only</option>
+            <option value="all" style="background-color: #FFFFFF; color: #000000;">All</option>
```
Comment (AlkTheOrg) [inline popup.html]: I thought we decided to revert all color changes, but you have bunch of inline stylings here. Also we have bootstrap, so why do you need inline styles for white and black?

File: getMultimedias.js (hunk 6 above)
Comment (AMMARmms) [reply getMultimedias.js]: Yes, like this one here.

ScreenShot_20240603215724

File: getMultimedias.js (hunk 1 above)
Comment (AMMARmms) [reply getMultimedias.js]: ok, as u like, it's your code.

File: popup.js (hunk 2 above)
Comment (AMMARmms) [reply popup.js]: ok, u are right here.

File: getMultimedias.js (hunk 7 above)
Comment (AMMARmms) [reply getMultimedias.js]: ok

File: getMultimedias.js (hunk 8 above)
Comment (AMMARmms) [reply getMultimedias.js]: ok

File: popup.css (hunk 9 above)
Comment (AMMARmms) [reply popup.css]: Sorry, old habits. it's not needed.

File: popup.css (hunk 12 above)
Comment (AMMARmms) [reply popup.css]: plz, provide the hex of the one u want.

File: popup.html (hunk 14 above)
Comment (AMMARmms) [reply popup.html]: noticed

File: popup.html (hunk 13 above)
Comment (AMMARmms) [reply popup.html]: noticed

File: popup.html (hunk 15 above)
Comment (AMMARmms) [reply popup.html]: Sorry, old habits. I just forgot to revert it.

File: getMultimedias.js (hunk 1 above)
Comment (AMMARmms) [reply getMultimedias.js]: But, the good thing about my code that it keeps the viewport above this black footer while scrolling 

ScreenShot_20240603221826

 as it's really annoying because it prevents the user from seeing where the scrolling reached at the moment. With this black footer, The user won't know how far or near the scrolling is from the end until it reaches the end.

Also, this code shouldn't be affected by sizing,  I've tried it on all fonts and it's been fine.

Comment (AlkTheOrg) [reply getMultimedias.js]: I agree, but `handleScroll` and the event listeners are not the proper solution to this problem. If you really want to solve it, here are 2 possible solutions that comes to my mind rn:

- You can conditionally check if `scrollAmount` + `document.documentElement.scrollTop` exceeding the `maxScrollPos` you calculated and then update the scroll position based on that.
- If `scrollAmount` + `document.documentElement.scrollTop` colliding with the footer's Y position, then you can just scroll to the last multimedia container instead of updating scroll position with a static value.

File: popup.html [hunk 16]
```diff
@@ -32,7 +31,7 @@
               target="_blank"
               href="https://github.com/AlkTheOrg/imdb-watchlist-randomizer/blob/main/ABOUT.md#about-the-extension"
               >About</a
-            >
+			>
```
Comment (AMMARmms) [inline popup.html:34]: in my text editor, this spacing not exit

File: getMultimedias.js (hunk 1 above)
Comment (AMMARmms) [reply getMultimedias.js]: > If scrollAmount + document.documentElement.scrollTop colliding with the footer's Y position, then you can just scroll to the last multimedia container instead of updating scroll position with a static value.

updated based on this.

File: popup.html (hunk 14 above)
Comment (AMMARmms) [reply popup.html]: in my text editor, this spacing doesn't exit

File: getMultimedias.js [hunk 17]
```diff
@@ -46,6 +47,7 @@
         genres: "", // TODO: Cut support for this as IMDB no longer displays genre content
         ratingIMDB,
         ratingMetac,
+		plot,
```
Comment (AlkTheOrg) [inline getMultimedias.js]:
```suggestion
        ratingMetac,
	plot,
```

File: getMultimedias.js [hunk 18]
```diff
@@ -58,24 +60,26 @@
 
     // sends the constructor multimedia objects
     const multimedias = constructMultimediasFromDivs(multimediaDivs);
-    console.log(multimedias)
-
+    console.log(multimedias);
     chrome.runtime.sendMessage({
       message: "all_multimedia",
-      payload: JSON.stringify(multimedias),
+      payload: JSON.stringify(multimedias),
```
Comment (AlkTheOrg) [inline getMultimedias.js]:
```suggestion
      payload: JSON.stringify(multimedias),
```

File: getMultimedias.js [hunk 19]
```diff
@@ -58,24 +60,26 @@
 
     // sends the constructor multimedia objects
     const multimedias = constructMultimediasFromDivs(multimediaDivs);
-    console.log(multimedias)
-
+    console.log(multimedias);
```
Comment (AlkTheOrg) [inline getMultimedias.js:63]: Could you readd the blank line next to this

File: getMultimedias.js [hunk 20]
```diff
@@ -58,24 +60,26 @@
...
-    const allTitles = document.querySelectorAll('.ipc-metadata-list-summary-item');
+	// const scrollAmount = document.documentElement.scrollHeight - document.documentElement.clientHeight;	
+    const allTitles = document.querySelectorAll(".ipc-metadata-list-summary-item");
     const targetElement = allTitles[allTitles.length - 1];
+	const rect = targetElement.getBoundingClientRect();
+	
+	const scrollAmount = 212 * 25; // each element's height is close to 212px and to make sure we load all the images we scroll 25 elements by 25
+	const nextScrollPos = document.documentElement.scrollTop + scrollAmount;
+	if (nextScrollPos >= rect.top) targetElement.scrollIntoView({ behavior: 'smooth', block: 'center' });
+    else document.documentElement.scrollTop += scrollAmount;
```
Comment (AlkTheOrg) [inline getMultimedias.js]:
```suggestion
    const targetElement = allTitles[allTitles.length - 1];
    const targetElementTop = targetElement.getBoundingClientRect().top;
	
    const scrollAmount = 212 * 25; // each element's height is close to 212px and to make sure we load all the images we scroll 25 elements by 25
    const nextScrollPosition = document.documentElement.scrollTop + scrollAmount;

    if (nextScrollPosition >= targetElementTop)
      targetElement.scrollIntoView({ behavior: 'smooth', block: 'center' });
    else
      document.documentElement.scrollTop += scrollAmount;
```

File: popup.css [hunk 21]
```diff
@@ -101,7 +101,7 @@ border-radius: 5px;
     position: relative;
     min-height: 40px;
   }
-
+
```
Comment (AlkTheOrg) [inline popup.css]:
```suggestion

```

File: popup.html (hunk 16 above)
Comment (AlkTheOrg) [inline popup.html]:
```suggestion

```

File: popup.html [hunk 22]
```diff
@@ -96,6 +95,15 @@
             id="showPoster"
           />
           <label class="form-check-label" for="showPoster">Poster</label>
+        </div>
+		 <div class="form-check form-check-inline">
+          <input
```
Comment (AlkTheOrg) [inline popup.html]:
```suggestion
        </div>
        <div class="form-check form-check-inline">
          <input
```

File: popup.html [hunk 23]
```diff
@@ -96,6 +95,15 @@
...
+            checked
+            class="form-check-input"
+            type="checkbox"
+            id="showplot"
+          />
+          <label class="form-check-label" for="showplot">Plot</label>
```
Comment (AlkTheOrg) [inline popup.html]:
```suggestion
            id="showPlot"
          />
          <label class="form-check-label" for="showPlot">Plot</label>
```

File: popup.js [hunk 24]
```diff
@@ -6,6 +6,7 @@ const defaultOptions = {
   showGenres: false,
   showPoster: true,
   showRatings: true,
+  showplot: true,
```
Comment (AlkTheOrg) [inline popup.js]:
```suggestion
  showPlot: true,
```

File: popup.js [hunk 25]
```diff
@@ -106,6 +107,13 @@ const createGenresElem = (genres) => {
   return genresList;
 };
 
+const createplotElem = (plot) => {
+	const plotElem = document.createElement("ul");
+	plotElem.textContent = plot;
+    addClassesToHTMLElem(plotElem, "mb-2 plot list-inline");
+    return plotElem;	
+};
```
Comment (AlkTheOrg) [inline popup.js]:
```suggestion
const createPlotElem = (plot) => {
  const plotElem = document.createElement("ul");
  plotElem.textContent = plot;
  addClassesToHTMLElem(plotElem, "mb-2 plot list-inline");
  return plotElem;	
};
```

File: popup.js [hunk 26]
```diff
@@ -121,11 +129,12 @@ const createCreditsElem = (credits) => {
 const createDivFromMultimedia = (mmedia) => {
   const div = document.createElement("div");
   addClassesToHTMLElem(div, "randomized-mmedia w-75 text-center mx-auto p-2 pb-3");
-  const { showCredits, showGenres, showPoster, showRatings } = loadOptions;
+  const { showCredits, showGenres, showPoster, showRatings, showplot } = loadOptions;
```
Comment (AlkTheOrg) [inline popup.js]:
```suggestion
  const { showCredits, showGenres, showPoster, showRatings, showPlot } = loadOptions;
```

File: popup.js [hunk 27]
```diff
@@ -121,11 +129,12 @@ const createCreditsElem = (credits) => {
...
 
   div.appendChild(createTitleElem(mmedia));
   div.appendChild(createYearElem(mmedia.year));
   div.appendChild(createHeroicContent(mmedia, showPoster, showRatings));
+  showplot && div.appendChild(createplotElem(mmedia.plot));
```
Comment (AlkTheOrg) [inline popup.js]:
```suggestion
  showPlot && div.appendChild(createPlotElem(mmedia.plot));
```

File: popup.js [hunk 28]
```diff
@@ -163,13 +172,15 @@ const loadSyncedOptions = async () => {
 const constructOptionsFromForm = () => {
   const showRatings = document.querySelector("#showRatings").checked;
   const showPoster = document.querySelector("#showPoster").checked;
+  const showplot = document.querySelector("#showplot").checked;
```
Comment (AlkTheOrg) [inline popup.js]:
```suggestion
  const showPlot = document.querySelector("#showPlot").checked;
```

File: popup.js [hunk 29]
```diff
@@ -163,13 +172,15 @@ const loadSyncedOptions = async () => {
...
   const showGenres = document.querySelector("#showGenres").checked;
   const showCredits = document.querySelector("#showCredits").checked;
   const minIMDBRating = document.querySelector("#minIMDBRating").value;
   const multimediaType = document.querySelector("#multimediaType").value;
   return {
     showRatings,
     showPoster,
+	showplot,
     showGenres,
```
Comment (AlkTheOrg) [inline popup.js]:
```suggestion
    showPoster,
    showPlot,
    showGenres,
```

File: popup.js [hunk 30]
```diff
@@ -191,7 +202,7 @@ const saveOptionsToSync = () => {
   );
 };
 
-inputIds = ['showRatings', 'showPoster', 'showGenres', 'showCredits', 'minIMDBRating', 'multimediaType']
+inputIds = ['showRatings', 'showPoster', 'showGenres', 'showplot', 'showCredits', 'minIMDBRating', 'multimediaType']
```
Comment (AlkTheOrg) [inline popup.js]:
```suggestion
inputIds = ['showRatings', 'showPoster', 'showGenres', 'showPlot', 'showCredits', 'minIMDBRating', 'multimediaType']
```

**Review by AlkTheOrg — COMMENTED** [review]
Looks good!
Just added some lint and wording suggestions

File: popup.css (hunk 12 above)
Comment (AlkTheOrg) [reply popup.css]: Feel free to pick something that doesn't break the accessibility and doesn't look off with rest of the colors. I can update it later if I find a more suitable color.

Also @AMMARmms you can just apply these suggestions from GitHub. You will see `Add suggestion to batch` and then you can commit multiple suggestions at once. Then you can simply pull the new changes to your local and try them out.

File: getMultimedias.js (hunk 19 above)
Comment (AMMARmms) [reply getMultimedias.js:63]: Added

**Comment by AMMARmms** [comment]
I applied your suggestion, please let me know if you have any other comments.

**Comment by AlkTheOrg** [comment]
Thanks for you contribution!

**Comment by AMMARmms** [comment]
"My pleasure!"
